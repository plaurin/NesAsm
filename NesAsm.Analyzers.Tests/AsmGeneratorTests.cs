using ApprovalTests.Reporters;

namespace NesAsm.Analyzers.Tests;

[UseReporter(typeof(VisualStudioReporter))]
public class AsmGeneratorTests
{
    [Fact]
    public void TestExampleScript()
    {
        var generator = new AsmGenerator();
        var content = File.ReadAllText($@"..\..\..\..\NesAsm.Example\HiloWorld.cs");

        GeneratorTestUtilities.TestGenerator(generator, content);
    }

    [Fact]
    public void TestSimpleScript()
    {
        TestScript("SimpleScript.cs");
    }

    [Fact]
    public void TestMultiProcScript()
    {
        TestScript("MultiProcScript.cs");
    }

    [Fact]
    public void TestJumpSubroutineScript()
    {
        TestScript("JumpSubroutineScript.cs");
    }

    [Fact]
    public void TestCharsScript()
    {
        TestScript("CharsScript.cs");
    }

    [Fact]
    public void TestBranchScript()
    {
        TestScript("BranchScript.cs");
    }

    [Fact]
    public void TestParametersScript()
    {
        TestScript("ParametersScript.cs");
    }

    [Fact]
    public void TestSubroutineResultScript()
    {
        TestScript("SubroutineResultScript.cs");
    }

    [Fact]
    public void TestSubroutineResultsScript()
    {
        TestScript("SubroutineResultsScript.cs");
    }

    [Fact]
    public void TestCallingOtherScript()
    {
        TestScript("CallingOtherScript.cs");
    }

    [Fact]
    public void TestDataScript()
    {
        TestScript("DataScript.cs");
    }

    [Fact]
    public void TestAllInstructionsScript()
    {
        TestScript("AllInstructionsScript.cs");
    }

    private static void TestScript(string filename)
    {
        var generator = new AsmGenerator();
        var content = File.ReadAllText($@"..\..\..\TestFiles\{filename}");

        GeneratorTestUtilities.TestGenerator(generator, content);
    }
}
