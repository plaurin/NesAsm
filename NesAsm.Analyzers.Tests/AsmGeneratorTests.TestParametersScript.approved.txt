- Main diagnostics:
(28,34): error NA0001: Type long is not supported for method parameters
(30,9): error NA0004: Parameter type LDAa not supported (should be byte or bool)
(35,9): error NA0003: OpCode STA not supported to use parameters (should only be LDAa, LDXa or LDYa)
(39,23): error NA0001: Type long is not supported for method parameters
- SyntaxTrees Count: 2
- Output diagnostics:
- RunResult.GeneratedTrees Count: 1
- RunResults diagnostics:
(28,34): error NA0001: Type long is not supported for method parameters
(30,9): error NA0004: Parameter type LDAa not supported (should be byte or bool)
(35,9): error NA0003: OpCode STA not supported to use parameters (should only be LDAa, LDXa or LDYa)
(39,23): error NA0001: Type long is not supported for method parameters
- RunResult[0].Generator: Microsoft.CodeAnalysis.IncrementalGeneratorWrapper
- RunResult[0] diagnostics:
(28,34): error NA0001: Type long is not supported for method parameters
(30,9): error NA0004: Parameter type LDAa not supported (should be byte or bool)
(35,9): error NA0003: OpCode STA not supported to use parameters (should only be LDAa, LDXa or LDYa)
(39,23): error NA0001: Type long is not supported for method parameters
- RunResult[0].GeneratedSources Count: 1
- RunResult[0].Exception: 
- RunResult[0].GeneratedSources[0].HintName: Asm.ParametersScript.cs
- RunResult[0].GeneratedSources[0].SourceText:
/* Generator Asm code in file Asm.ParametersScript.s
; Auto generated code using the NesAsm project
.segment "CODE"

.proc main
  lda #25
  sta 0
  lda #232
  sta 1
  lda #3
  sta 2
  lda #1
  sta 3
  jsr procB

  rts
.endproc

.proc procB
  lda 0
  sta $40

  ; LDXa(b);
  ; STA(0x41);

  ldy 2
  sta $42

  rts
.endproc

.proc invalidParamType

  rts
.endproc

.proc invalidOpCodeUsingParameter

  rts
.endproc

.proc lDAa

  rts
.endproc


*/
